plugin-maven/src/test/java/com/diffplug/gradle/spotless/MavenRunner.java

```diff
@@ -0,0 +1,155 @@
+		return this;
+	}
+
+	public MavenRunner withArguments(String... args) {
+		this.args = Objects.requireNonNull(args);
+		return this;
+	}
+
+	private Result run() throws IOException, InterruptedException {
+		Objects.requireNonNull(projectDir, "Need to call withProjectDir() first");
+		Objects.requireNonNull(args, "Need to call withArguments() first");
+		// run maven with the given args in the given directory
+		List<String> cmds = getPlatformCmds("mvn " + Arrays.stream(args).collect(Collectors.joining(" ")));
+		ProcessBuilder builder = new ProcessBuilder(cmds);
+		builder.directory(projectDir);
+		Process process = builder.start();
+		// slurp and return the stdout, stderr, and exitValue
+		Slurper output = new Slurper(process.getInputStream());
+		Slurper error = new Slurper(process.getErrorStream());
+		int exitValue = process.waitFor();
+		output.join();
+		error.join();
+		return new Result(exitValue, output.result(), error.result());
+	}
+
+	/** Runs the command and asserts that nothing was printed to stderr. */
+	public Result runNoError() throws IOException, InterruptedException {
+		Result result = run();
+		Assertions.assertThat(result.error()).isEmpty();
+		return result;
+	}
+
+	/** Runs the command and asserts that something was printed to stderr. */
+	public Result runHasError() throws IOException, InterruptedException {
+		Result result = run();
+		Assertions.assertThat(result.error()).isNotEmpty();
+		return result;
+	}
+
+	public static class Result {
+		private final int exitValue;
+		private final String output;
+		private final String error;
+
+		public Result(int exitValue, String output, String error) {
+			super();
+			this.exitValue = exitValue;
+			this.output = Objects.requireNonNull(output);
+			this.error = Objects.requireNonNull(error);
+		}
+
+		public int exitValue() {
+			return exitValue;
+		}
+
+		public String output() {
+			return output;
+		}
+
+		public String error() {
+			return error;
+		}
+	}
+
+	/** Prepends any arguments necessary to run a console command. */
+	private static List<String> getPlatformCmds(String cmd) {
+		if (isWin()) {
+			return Arrays.asList("cmd", "/c", cmd);
+		} else {
+			return Arrays.asList("/bin/sh", "-c", cmd);
+		}
+	}
+
+	private static boolean isWin() {
+		String os_name = System.getProperty("os.name").toLowerCase(Locale.getDefault());
+		return os_name.contains("win");
+	}
+
+	private static class Slurper extends Thread {
+		private final InputStream input;
+		private volatile String result;
+
+		Slurper(InputStream input) {
+			this.input = Objects.requireNonNull(input);
+			start();
+		}
+
+		@Override
+		public void run() {
+			try {
+				ByteArrayOutputStream output = new ByteArrayOutputStream();
+				byte[] buffer = new byte[1024];
+				int numRead;
+				while ((numRead = input.read(buffer)) != -1) {
+					output.write(buffer, 0, numRead);
+				}
+				result = output.toString(Charset.defaultCharset().name());
```

Comment:
@nedtwigg Minor nit: I seem to remember that we have access to Guava in some way through a re-packaged version that Diffplug distributes on Maven Central and JCenter.

If we do have access to Guava in this way, then I think we should consider using [`ByteStreams.copy`](https://static.javadoc.io/com.google.guava/guava/23.6-jre/com/google/common/io/ByteStreams.html#copy-java.io.InputStream-java.io.OutputStream-) to simplify the code inside this try statement. But I'd understand if you don't want to do this for dependency reasons or such. :)

Reply:
Good call, implemented in next commit.  For the rest of this review, I'd like to focus on bashing out the high-level design.  We can fix performance nits and duplicate code after we have something working :)

Reply:
Roger! :)